docs/_docs/02_testing-best-practices/unit-integration-end-to-end-test.md

```diff
@@ -22,3 +22,149 @@ Link to the video at [infoq.com](https://www.infoq.com/presentations/automated-t
 ## Slides
 
 Slides to the video can be found here: [Slides: How to test infrastructure code](https://www.slideshare.net/brikis98/how-to-test-infrastructure-code-automated-testing-for-terraform-kubernetes-docker-packer-and-more){:target="_blank"}.
+
+## Unit test helpers
+
+The function terraform.UnitTest will help unit test a single module, the UnitTest function will do the following things
+
+**_TODO_**: As this is a one day project I will only attempt to do step 3, 4, and 6 today. Time permitting I will add 2
+as well 
+
+1. Copy your terraform code to a temporary folder
+2. Optionally call a setup function you provide
+3. Init and Apply the terraform code
+4. Call a list of validators you provide
+5. Optionally call a tearDown function you provide
+6. Destroy the terraform code
+
+Most of these steps are wrapped into stages which you can skip (see [Iterating locally using test stages]({% link _docs/02_testing-best-practices/iterating-locally-using-test-stages.md %})). This means you can skip individual stages
+by setting env var SKIP_\<stage_name\> to any value.
+
+If you skip the init_apply step then UnitTest will automatically load your previously applied outputs from the working directory.
+
+The full list of stages is:
+
+* setup
+* init_apply
+* validate
+* teardown
+* destroy
+
+Signature of UnitTest:
+
+```go
+// The output of the module after it has been applied, as returned by terraform.OutputAll
+type ModuleOutputs map[string]interface{}
+
+// A validation function to test your terraform code
+type Validator func(t *testing.T, moduleOutputs *terraform.ModuleOutputs, workingDir string)
+
+type UnitTestPlan struct {
+  TerraformOptions *terraform.Options
+  // If you do not provide a working directory it will default to ./.terratest-unit-test/TEST_FUNCTION_NAME/

```

Comment:
Is this more of a scratch directory?

Reply:
No, since we are going to be automating saving and loading the terraform options (and maybe other things, like the testData) we have to use some working directory or other, we can allow the user to specify it, but so they don't have to (and so it could be easily git ignored) I chose this path. The test name comes from testing.T and means we can have a unique directory per test, so they wont collide even when run in parallel.

See the simple implementation here: https://github.com/jfharden/terratest/pull/1/files#diff-abf55a1c7dcad9df4f582c66ee96b4d8R32